File: pythonLeetcode/208.实现-trie-前缀树.py
#
# @lc app=leetcode.cn id=208 lang=python
#
# [208] 实现 Trie (前缀树)
#
# https://leetcode-cn.com/problems/implement-trie-prefix-tree/description/
#
# algorithms
# Medium (62.39%)
# Likes:    163
# Dislikes: 0
# Total Accepted:    18.6K
# Total Submissions: 29.4K
# Testcase Example:  '["Trie","insert","search","search","startsWith","insert","search"]\n' +
  '[[],["apple"],["apple"],["app"],["app"],["app"],["app"]]'
#
# 实现一个 Trie (前缀树)，包含 insert, search, 和 startsWith 这三个操作。
# 
# 示例:
# 
# Trie trie = new Trie();
# 
# trie.insert("apple");
# trie.search("apple");   // 返回 true
# trie.search("app");     // 返回 false
# trie.startsWith("app"); // 返回 true
# trie.insert("app");   
# trie.search("app");     // 返回 true
# 
# 说明:
# 
# 
# 你可以假设所有的输入都是由小写字母 a-z 构成的。
# 保证所有输入均为非空字符串。
# 
# 
#

# @lc code=start

# 15/15 cases passed (136 ms)
# Your runtime beats 79.7 % of python submissions
# Your memory usage beats 57.01 % of python submissions (27.8 MB)
class Trie(object):

    def __init__(self):
        self.root = {}


    def insert(self, word):
        tmpTrie = self.root
        for itemChar in word:
            if itemChar not in tmpTrie:
                tmpTrie[itemChar] = {}
            tmpTrie = tmpTrie[itemChar]
        tmpTrie['#'] = '#'

        

    def search(self, word):
        tmpTrie = self.root
        for itemChar in word:
            if itemChar not in tmpTrie:
                return False 
            tmpTrie = tmpTrie[itemChar]
        if '#' in tmpTrie:
            return True
        return False        

    def startsWith(self, prefix):
        tmpTrie = self.root
        for itemChar in prefix:
            if itemChar not in tmpTrie:
                return False 
            tmpTrie = tmpTrie[itemChar]

            
        return True

# @lc code=end

File: LoveSwiftLeetcode/208.实现-trie-前缀树.swift
/*
 * @lc app=leetcode.cn id=208 lang=swift
 *
 * [208] 实现 Trie (前缀树)
 */

// @lc code=start

class Trie {
    class TrieTreeNode {
        var children: [Character: TrieTreeNode] = [:]
        var isWord = false
    }
    let root = TrieTreeNode()
    init() {

    }
    
    func insert(_ word: String) {
        var cur = root
        for c in word {
            if let cn = cur.children[c] {
                cur = cn
            } else {
                cur.children[c] = TrieTreeNode()
                cur = cur.children[c]!
            }
        }
        cur.isWord = true
    }
    
    func search(_ word: String) -> Bool {
        var cur = root
        for c in word {
            if let cn = cur.children[c] {
                cur = cn
            } else {
                return false
            }
        }
        return cur.isWord
        
    }
    
    func startsWith(_ word: String) -> Bool {
        var cur = root
        for c in word {
            if let cn = cur.children[c] {
                cur = cn
            } else {
                return false
            }
        }
        return true
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * let obj = Trie()
 * obj.insert(word)
 * let ret_2: Bool = obj.search(word)
 * let ret_3: Bool = obj.startsWith(prefix)
 */
// @lc code=end

